File: components/mardown-display/blocks/quiz/quiz-parser.ts
/**
 * Quiz Parser and Hash Utilities
 * Handles parsing quiz JSON with content hashing
 */

import type { OriginalQuestion } from './quiz-types';

export type QuizData = {
  questions: OriginalQuestion[];
  title: string;
  category?: string;
  contentHash: string;
};

export type RawQuizJSON = {
  quizTitle: string;
  category?: string;
  multipleChoice: Array<{
    id: number;
    question: string;
    options: string[];
    correctAnswer: number;
    explanation: string;
  }>;
};

/**
 * Generate a stable hash from quiz content
 * Uses only questions to create hash (ignoring metadata like title)
 */
export async function generateQuizHash(questions: OriginalQuestion[]): Promise<string> {
  // Create a stable string representation of questions
  // Sort by ID to ensure consistency
  const sortedQuestions = [...questions].sort((a, b) => a.id - b.id);
  
  const contentString = JSON.stringify(sortedQuestions.map(q => ({
    question: q.question.trim().toLowerCase(),
    options: q.options.map(o => o.trim().toLowerCase()).sort(),
    correctAnswer: q.correctAnswer
  })));

  // Generate SHA-256 hash
  const encoder = new TextEncoder();
  const data = encoder.encode(contentString);
  const hashBuffer = await crypto.subtle.digest('SHA-256', data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  const hashHex = hashArray.map(b => b.toString(16).padStart(2, '0')).join('');
  
  return hashHex;
}

/**
 * Parse quiz JSON
 */
export async function parseQuizJSON(jsonData: RawQuizJSON): Promise<QuizData> {
  const questions = [...jsonData.multipleChoice];
  questions.sort((a, b) => a.id - b.id);
  const contentHash = await generateQuizHash(questions);
  return {
    questions,
    title: jsonData.quizTitle,
    category: jsonData.category,
    contentHash
  };
}

/**
 * Parse quiz from string (convenience wrapper)
 */
export async function parseQuizString(jsonString: string): Promise<QuizData> {
  const parsed = JSON.parse(jsonString);
  return parseQuizJSON(parsed);
}

/**
 * Check if two quiz hashes match
 */
export function quizHashesMatch(hash1: string, hash2: string): boolean {
  return hash1.toLowerCase() === hash2.toLowerCase();
}

/**
 * Validate quiz data has required fields
 */
export function isValidQuizData(data: any): data is RawQuizJSON {
  if (!data || typeof data !== 'object') return false;
  if (!data.quizTitle || typeof data.quizTitle !== 'string') return false;
  if (!Array.isArray(data.multipleChoice) || data.multipleChoice.length === 0) return false;
  return data.multipleChoice.every((q: any) =>
    typeof q.id === 'number' &&
    typeof q.question === 'string' &&
    Array.isArray(q.options) &&
    q.options.length > 0 &&
    typeof q.correctAnswer === 'number' &&
    typeof q.explanation === 'string'
  );
}

File: lib/python/parsers/quiz_parser.py
"""Quiz parser — port of quiz-parser.ts."""

from __future__ import annotations

import hashlib
import json

from lib.python.models.quiz import QuizBlockData, QuizQuestion


def parse_quiz(content: str) -> QuizBlockData | None:
    """Parse quiz JSON string into structured data."""
    try:
        data = json.loads(content.strip())
    except json.JSONDecodeError:
        return None

    if not _is_valid_quiz(data):
        return None

    questions = sorted(data["multiple_choice"], key=lambda q: q.get("id", 0))
    content_hash = _generate_quiz_hash(questions)

    return QuizBlockData(
        title=data["quiz_title"],
        category=data.get("category"),
        questions=[
            QuizQuestion(
                id=q["id"],
                question=q["question"],
                options=q["options"],
                correct_answer=q["correctAnswer"],
                explanation=q["explanation"],
            )
            for q in questions
        ],
        content_hash=content_hash,
    )


def _is_valid_quiz(data: dict) -> bool:
    if not isinstance(data, dict):
        return False
    if not data.get("quiz_title") or not isinstance(data.get("quiz_title"), str):
        return False
    mc = data.get("multiple_choice")
    if not isinstance(mc, list) or len(mc) == 0:
        return False
    return all(
        isinstance(q.get("id"), (int, float))
        and isinstance(q.get("question"), str)
        and isinstance(q.get("options"), list)
        and len(q["options"]) > 0
        and isinstance(q.get("correctAnswer"), (int, float))
        and isinstance(q.get("explanation"), str)
        for q in mc
    )


def _generate_quiz_hash(questions: list[dict]) -> str:
    sorted_qs = sorted(questions, key=lambda q: q.get("id", 0))
    normalized = json.dumps(
        [
            {
                "question": q["question"].strip().lower(),
                "options": sorted(o.strip().lower() for o in q["options"]),
                "correctAnswer": q["correctAnswer"],
            }
            for q in sorted_qs
        ],
        sort_keys=True,
    )
    return hashlib.sha256(normalized.encode()).hexdigest()
